Wrappers/Python/cil/processors/Binner.py

```diff
@@ -14,218 +14,410 @@
-                    step = 1
-                    if roi[label][0] != None:
-                        if roi[label][0] < 0:
-                            stop =  n_elements[i]+roi[label][0]
-                        else:
-                            stop = roi[label][0]
-                    else:
-                        stop =  n_elements[i]
-
-                elif len(roi[label]) == 2 or len(roi[label]) == 3:
-                    
-                    if roi[label][0] != None:
-                        if roi[label][0] < 0:
-                            start = n_elements[i]+roi[label][0]
-                        else:
-                            start = roi[label][0]
-                    else:
-                        start = 0
-
-                    if roi[label][1] != None:
-                        if roi[label][1] < 0:
-                            stop = n_elements[i]+roi[label][1]
-                        else:
-                            stop = roi[label][1]
-                    else: 
-                        stop = n_elements[i]
-                    
-                    if len(roi[label]) == 2:
-                        step = 1
-
-                    if len(roi[label]) == 3:
-                        if roi[label][2] != None:
-                            if roi[label][2] <= 0:
-                                raise ValueError('Binning parameter has to be > 0')
-                            else:
-                                step = roi[label][2]
-                        else:
-                            step = 1
-                else:
-                    raise ValueError('roi is exected to have 1, 2 or 3 elements')
+        # create output array or check size and shape of passed out
+        if out is None:
+            if binned_geometry is not None:
+                data_out = binned_geometry.allocate(None)
+                binned_array = data_out.as_array()
             else:
-                step = 1
-                start = 0
-                stop = n_elements[i]
-            roi_object.append((start, stop, step))
-        return roi_object
\ No newline at end of file
+                binned_array = np.empty(self._shape_out,dtype=np.float32)
+                data_out = DataContainer(binned_array,False, data.dimension_labels)
+        else:
+            try:
+                out.array = np.asarray(out.array, dtype=np.float32, order='C').reshape(self._shape_out)
+            except:
+                raise ValueError("Array of `out` not compatible. Expected shape: {0}, data type: {1} Got shape: {2}, data type: {3}".format(self._shape_out, np.float32, out.array.shape, out.array.dtype))
+
+            if binned_geometry is not None:
+                if out.geometry != binned_geometry:
+                    raise ValueError("Geometry of `out` not as expected. Got {0}, expected {1}".format(out.geometry, binned_geometry))
+            
+            binned_array = out.array
+
+        # bin data
+        if self._accelerated:
```

Comment:
I wonder what happens if the user doesn't have IPP and doesn't change the default. It'll fail? We happy with that?

Reply:
runtime error is raised :)